Add PartitonHub, #21880

patriknw opened the pull request:
For early review and discussion.

Refs #21880

patriknw committed:
Add PartitonHub, #21880

* FixedSizePartitionHub in Artery
* expose consumer queue size

patriknw commented on akka-bench-jmh/src/main/scala/akka/stream/PartitionHubBenchmark.scala:
```diff
@@ -0,0 +1,125 @@
+import akka.stream.impl.StreamSupervisor
+import akka.stream.scaladsl.PartitionHub
+
+object PartitionHubBenchmark {
```
```
jmh:run -i 10 -wi 5 -f2 -t1 akka.stream.PartitionHubBenchmark

[info] Benchmark                        (BufferSize)  (NumberOfStreams)   Mode  Cnt        Score        Error  Units
[info] PartitionHubBenchmark.partition           256                  1  thrpt   20  2975068.781 ± 134146.968  ops/s
[info] PartitionHubBenchmark.partition           256                  5  thrpt   20  4156257.425 ± 306498.950  ops/s
[info] PartitionHubBenchmark.partition           256                 10  thrpt   20  3866510.802 ± 378733.234  ops/s
```

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -693,3 +697,385 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+   *                   concurrent consumers can be in terms of element. If this buffer is full, the producer
+   *                   is backpressured. Must be a power of two and less than 4096.
+   */
+  def sink[T](partitioner: T ⇒ Int, startAfterNbrOfStreams: Int, bufferSize: Int = 256): Sink[T, Source[T, NotUsed]] =
```
There are several things around the api we should discuss.

* does the `partitioner` function need to know how many streams that are currently registered?
* should the `partitioner` function decide if the element should be buffered in pending instead of having the `startAfterNbrOfStreams` parameter?
* should the `startAfterNbrOfStreams` also be used after it has been started to run and a stream is deregistered, so it's a `minNbrOfStreams`?

patriknw replied:
After thinking some about this I think it's best to include number of streams in the function.
```
partition: (T, Int) => Int
```

`startAfterNbrOfStreams` can still be a separate parameter

johanandren replied:
I was even thinking of something to identify each downstream, but maybe that would be a different thing, something like a list of unique id's, so it would be possible to implement stickyness, once you started sending T to a specific downstream you could try keep doing that even as long as that specific downstream still is running.

ktoso replied:
The `(T, Int) => Int` paritioner function looks good.

I'm not sure I get the additional 2 questions, could you clarify them a bit?

patriknw replied:
`(T, Int) => Int` works for the hashCode modulo use case, but I think there can be other use cases.
1) sticky session, like Johan mention, e.g. route all elements for user "Bob" to a specific stream, until that stream disappears and then pick another
2) round robin for each element, also requires state (the counter) in the function as another delicate

In the implementation each stream has a `Long` id (it's just a counter). I'm thinking of exposing those as `Array[Long] => Long`. Then you know the size, and could keep track of a sticky session mapping. To support state it has to be `() => (Array[Long] => Long)`, which is always a bit mind boggling.

We could definitely have a simple and a more powerful API. We could also provide pre-packaged functions, such as for the round-robin case.

patriknw replied:
Regarding `startAfterNbrOfStreams` is for when you don't want to route any elements until enough number of streams have been registered. E.g. in Artery the number of streams is fixed, and known, and messages can't be routed until all streams are there.

patriknw commented:
I have optimized it some more and use it in Artery.

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -693,3 +699,423 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+
+    // Need the queue to be pluggable to be able to use a more performant (less general)
+    // queue in Artery
+    trait PartitionQueue {
```
Here are the benchmark results:
```
[info] Benchmark                                                           (BufferSize)  (NumberOfStreams)   Mode  Cnt        Score        Error   Units
[info] PartitionHubBenchmark.partition                                              256                  1  thrpt   12  4472452.035 ± 379272.863   ops/s
[info] PartitionHubBenchmark.partition:·gc.alloc.rate                               256                  1  thrpt   12     1214.771 ±    329.219  MB/sec
[info] PartitionHubBenchmark.partition:·gc.alloc.rate.norm                          256                  1  thrpt   12      287.892 ±     87.997    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space                      256                  1  thrpt   12     1256.202 ±    434.457  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space.norm                 256                  1  thrpt   12      297.412 ±    108.877    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space                  256                  1  thrpt   12        0.021 ±      0.019  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space.norm             256                  1  thrpt   12        0.005 ±      0.005    B/op
[info] PartitionHubBenchmark.partition:·gc.count                                    256                  1  thrpt   12       28.000               counts
[info] PartitionHubBenchmark.partition:·gc.time                                     256                  1  thrpt   12       23.000                   ms
[info] PartitionHubBenchmark.partition                                              256                  5  thrpt   12  4758452.104 ± 529411.579   ops/s
[info] PartitionHubBenchmark.partition:·gc.alloc.rate                               256                  5  thrpt   12      471.910 ±     48.670  MB/sec
[info] PartitionHubBenchmark.partition:·gc.alloc.rate.norm                          256                  5  thrpt   12      104.415 ±      0.849    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space                      256                  5  thrpt   12      469.124 ±     91.323  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space.norm                 256                  5  thrpt   12      103.850 ±     18.231    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space                  256                  5  thrpt   12        0.041 ±      0.031  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space.norm             256                  5  thrpt   12        0.009 ±      0.007    B/op
[info] PartitionHubBenchmark.partition:·gc.count                                    256                  5  thrpt   12       34.000               counts
[info] PartitionHubBenchmark.partition:·gc.time                                     256                  5  thrpt   12       20.000                   ms
[info] PartitionHubBenchmark.partition                                              256                 10  thrpt   12  4529333.257 ±  62294.979   ops/s
[info] PartitionHubBenchmark.partition:·gc.alloc.rate                               256                 10  thrpt   12      460.360 ±      5.455  MB/sec
[info] PartitionHubBenchmark.partition:·gc.alloc.rate.norm                          256                 10  thrpt   12      107.004 ±      0.127    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space                      256                 10  thrpt   12      469.674 ±     87.107  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space.norm                 256                 10  thrpt   12      109.167 ±     20.081    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space                  256                 10  thrpt   12        0.039 ±      0.034  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space.norm             256                 10  thrpt   12        0.009 ±      0.008    B/op
[info] PartitionHubBenchmark.partition:·gc.count                                    256                 10  thrpt   12       33.000               counts
[info] PartitionHubBenchmark.partition:·gc.time                                     256                 10  thrpt   12       21.000                   ms
```

and compared to when the `ConcurrentHashMap` is used (when id >= 128):
```
[info] Benchmark                                                           (BufferSize)  (NumberOfStreams)   Mode  Cnt        Score        Error   Units
[info] PartitionHubBenchmark.partition                                              256                  1  thrpt   12   991680.354 ± 100727.118   ops/s
[info] PartitionHubBenchmark.partition:·gc.alloc.rate                               256                  1  thrpt   12     1614.999 ±    219.292  MB/sec
[info] PartitionHubBenchmark.partition:·gc.alloc.rate.norm                          256                  1  thrpt   12     1710.444 ±     76.052    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space                      256                  1  thrpt   12     1651.561 ±    270.543  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space.norm                 256                  1  thrpt   12     1749.595 ±    190.958    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space                  256                  1  thrpt   12        0.052 ±      0.085  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space.norm             256                  1  thrpt   12        0.055 ±      0.092    B/op
[info] PartitionHubBenchmark.partition:·gc.count                                    256                  1  thrpt   12       53.000               counts
[info] PartitionHubBenchmark.partition:·gc.time                                     256                  1  thrpt   12       61.000                   ms
[info] PartitionHubBenchmark.partition                                              256                  5  thrpt   12  2638442.541 ±  79513.192   ops/s
[info] PartitionHubBenchmark.partition:·gc.alloc.rate                               256                  5  thrpt   12     1006.254 ±     31.850  MB/sec
[info] PartitionHubBenchmark.partition:·gc.alloc.rate.norm                          256                  5  thrpt   12      401.768 ±     20.979    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space                      256                  5  thrpt   12     1043.761 ±    320.936  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space.norm                 256                  5  thrpt   12      417.359 ±    131.506    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space                  256                  5  thrpt   12        0.036 ±      0.055  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space.norm             256                  5  thrpt   12        0.014 ±      0.022    B/op
[info] PartitionHubBenchmark.partition:·gc.count                                    256                  5  thrpt   12       45.000               counts
[info] PartitionHubBenchmark.partition:·gc.time                                     256                  5  thrpt   12       33.000                   ms
[info] PartitionHubBenchmark.partition                                              256                 10  thrpt   12  4088010.755 ±  69644.716   ops/s
[info] PartitionHubBenchmark.partition:·gc.alloc.rate                               256                 10  thrpt   12      438.160 ±      5.455  MB/sec
[info] PartitionHubBenchmark.partition:·gc.alloc.rate.norm                          256                 10  thrpt   12      112.839 ±      0.706    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space                      256                 10  thrpt   12      437.113 ±     76.592  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space.norm                 256                 10  thrpt   12      112.608 ±     20.022    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space                  256                 10  thrpt   12        0.036 ±      0.037  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space.norm             256                 10  thrpt   12        0.009 ±      0.009    B/op
[info] PartitionHubBenchmark.partition:·gc.count                                    256                 10  thrpt   12       30.000               counts
[info] PartitionHubBenchmark.partition:·gc.time                                     256                 10  thrpt   12       18.000                   ms
```

and compared to the `FixedSizePartitionHub` that is used in Artery, especially note the lower allocation rate:
```
[info] Benchmark                                                           (BufferSize)  (NumberOfStreams)   Mode  Cnt        Score        Error   Units
[info] PartitionHubBenchmark.arteryLanes                                            256                  1  thrpt   12  6889256.962 ± 531278.945   ops/s
[info] PartitionHubBenchmark.arteryLanes:·gc.alloc.rate                             256                  1  thrpt   12        8.263 ±      2.803  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.alloc.rate.norm                        256                  1  thrpt   12        1.283 ±      0.534    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Eden_Space                    256                  1  thrpt   12       10.431 ±     31.206  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Eden_Space.norm               256                  1  thrpt   12        1.568 ±      4.697    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Survivor_Space                256                  1  thrpt   12        0.281 ±      0.841  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Survivor_Space.norm           256                  1  thrpt   12        0.042 ±      0.127    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.count                                  256                  1  thrpt   12        2.000               counts
[info] PartitionHubBenchmark.arteryLanes:·gc.time                                   256                  1  thrpt   12       10.000                   ms
[info] PartitionHubBenchmark.arteryLanes                                            256                  5  thrpt   12  6059917.035 ± 498690.819   ops/s
[info] PartitionHubBenchmark.arteryLanes:·gc.alloc.rate                             256                  5  thrpt   12       34.218 ±      2.471  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.alloc.rate.norm                        256                  5  thrpt   12        5.948 ±      0.233    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Eden_Space                    256                  5  thrpt   12       36.023 ±     40.771  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Eden_Space.norm               256                  5  thrpt   12        6.390 ±      7.268    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Survivor_Space                256                  5  thrpt   12        0.956 ±      2.787  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Survivor_Space.norm           256                  5  thrpt   12        0.167 ±      0.487    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.count                                  256                  5  thrpt   12        7.000               counts
[info] PartitionHubBenchmark.arteryLanes:·gc.time                                   256                  5  thrpt   12       44.000                   ms
[info] PartitionHubBenchmark.arteryLanes                                            256                 10  thrpt   12  5634594.361 ±  83081.304   ops/s
[info] PartitionHubBenchmark.arteryLanes:·gc.alloc.rate                             256                 10  thrpt   12       53.854 ±      1.184  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.alloc.rate.norm                        256                 10  thrpt   12       10.062 ±      0.221    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Eden_Space                    256                 10  thrpt   12       54.719 ±     22.260  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Eden_Space.norm               256                 10  thrpt   12       10.215 ±      4.159    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Survivor_Space                256                 10  thrpt   12        0.985 ±      2.856  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Survivor_Space.norm           256                 10  thrpt   12        0.185 ±      0.537    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.count                                  256                 10  thrpt   12       11.000               counts
[info] PartitionHubBenchmark.arteryLanes:·gc.time                                   256                 10  thrpt   12       34.000                   ms
```

ktoso replied:
I like the pluggable queue design, very nice.

ktoso replied:
Hah, I was comparing the first too and was worried not seeing much grains there, and only then noticed the fixed size numbers :-) Very nice :-)

patriknw commented:
Latency distributions for Artery old and new with 4 lanes at 10000 msg/s (and also 1 lane):
<img width="1206" alt="latency-10000" src="https://user-images.githubusercontent.com/336161/26916432-aad5e42c-4c2a-11e7-9d99-5446e266f7c6.png">

Latency distributions for old and new with 4 lanes at 20000 msg/s:

<img width="1218" alt="latency-20000a" src="https://user-images.githubusercontent.com/336161/26916502-eed14cf2-4c2a-11e7-8d40-4d9c2625fa61.png">

Removing the old, latency distributions for new with 4 lanes at 20000 msg/s:

<img width="1191" alt="latency-20000b" src="https://user-images.githubusercontent.com/336161/26916512-f5206ab6-4c2a-11e7-85b9-9aa42ab21a44.png">

patriknw commented:
Artery throughput with 4 lanes with new hub:
```
[JVM-1] === MaxThroughput 1-to-1-snd1: throughput 992,851 msg/s, 99,285,147 bytes/s (payload), 137,013,503 bytes/s (total,compression), dropped 0, max round-trip 35 ms, burst size 1000, payload size 100, total size 138, 10072 ms to deliver 10000000 messages.
[JVM-1] === MaxThroughput 1-to-1-size-1k-snd1: throughput 491,461 msg/s, 491,460,867 bytes/s (payload), 510,136,380 bytes/s (total,compression), dropped 0, max round-trip 54 ms, burst size 1000, payload size 1000, total size 1038, 8139 ms to deliver 4000000 messages.
[JVM-1] === MaxThroughput 5-to-5-snd3: throughput 632,571 msg/s, 63,257,109 bytes/s (payload), 87,294,810 bytes/s (total,compression), dropped 0, max round-trip 151 ms, burst size 200, payload size 100, total size 138, 31617 ms to deliver 4000000 messages.
```

compared to old hub with 4 lanes:
```
[JVM-1] === MaxThroughput 1-to-1-snd1: throughput 813,074 msg/s, 81,307,423 bytes/s (payload), 112,204,244 bytes/s (total,compression), dropped 0, max round-trip 50 ms, burst size 1000, payload size 100, total size 138, 12299 ms to deliver 10000000 messages.
[JVM-1] === MaxThroughput 1-to-1-size-1k-snd1: throughput 562,430 msg/s, 562,429,696 bytes/s (payload), 583,802,025 bytes/s (total,compression), dropped 0, max round-trip 47 ms, burst size 1000, payload size 1000, total size 1038, 7112 ms to deliver 4000000 messages.
[JVM-1] === MaxThroughput 5-to-5-snd3: throughput 618,774 msg/s, 61,877,359 bytes/s (payload), 85,390,756 bytes/s (total,compression), dropped 0, max round-trip 74 ms, burst size 200, payload size 100, total size 138, 32322 ms to deliver 4000000 messages.
```

and compared with 1 lane:
```
[JVM-1] === MaxThroughput 1-to-1-snd1: throughput 744,713 msg/s, 74,471,254 bytes/s (payload), 102,770,331 bytes/s (total,compression), dropped 0, max round-trip 66 ms, burst size 1000, payload size 100, total size 138, 13428 ms to deliver 10000000 messages.
[JVM-1] === MaxThroughput 1-to-1-size-1k-snd1: throughput 496,648 msg/s, 496,647,629 bytes/s (payload), 515,520,238 bytes/s (total,compression), dropped 0, max round-trip 96 ms, burst size 1000, payload size 1000, total size 1038, 8054 ms to deliver 4000000 messages.
[JVM-1] === MaxThroughput 5-to-5-snd2: throughput 633,894 msg/s, 63,389,433 bytes/s (payload), 87,477,418 bytes/s (total,compression), dropped 0, max round-trip 140 ms, burst size 200, payload size 100, total size 138, 31551 ms to deliver 4000000 messages.
```

johanandren commented:
Any idea why the 25% improvement in throughput is only for 1-to-1 and not the other two?
What are the `latency....-cpulevel10` numbers?

ktoso commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -693,3 +699,423 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+  val in: Inlet[T] = Inlet("PartitionHub.in")
+  override val shape: SinkShape[T] = SinkShape(in)
+
+  def createQueue(): PartitionQueue = new PartitionQueueImpl
```
add a small comment that this is intended to be overriden with specialized queues if needed

ktoso commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -693,3 +699,423 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+      failStage(ex)
+    }
+
+    override def postStop(): Unit = {
```
sanity check, termination of hub happens when upstream completes, then we notify all downstreams using the callbacks

ktoso commented:
Reviewed the the stage impls rather in detail and I think they look correct. Could you clarify the questions you had there though?

patriknw commented on akka-bench-jmh/src/main/scala/akka/remote/artery/LatchSink.scala:
```diff
@@ -24,6 +24,11 @@ class LatchSink(countDownAfter: Int, latch: CountDownLatch) extends GraphStage[S
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        println(ex.getMessage)
+        ex.printStackTrace()
+      }
```
this is in test code, and actually pretty useful

ktoso replied:
Ok, I missed that it's in tests :)

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala:
```diff
@@ -367,4 +367,231 @@ class HubSpec extends StreamSpec {
+      val result2 = source.runWith(Sink.seq)
+      result1.futureValue should ===(List("usr-2"))
+      result2.futureValue should ===(List("usr-1", "usr-1", "usr-3"))
+    }
```
I have added the more powerful API: `() => (Array[Long], T) => Long` and above tests illustrate how that can be used to implement stateful routing. Question is if we also need the more simple index based API `(Int, T) => Int`?

The happy case above would then be:
```
PartitionHub.simpleSink((size, elem) ⇒ elem % size)
```

or should we make it even more specialized for `hashCode` like scenarios: `T => Int`, e.g.
```
PartitionHub.hashingSink(elem ⇒ elem.hashCode)
```

johanandren replied:
Nice!

What about having the factories `sink((ids, elem) => id)` and `statefulSink(() => (ids, elem) => id)` to match `mapConcat` and `statefulMapConcat`? Or maybe this is enough and kind of a power user stage anyways so the stateful one could be the only one. On the fence.

ktoso replied:
For reference; Stateful things also handled in new API with ConsumerInfo

johanandren commented on akka-remote/src/main/scala/akka/remote/artery/FixedSizePartitionHub.scala:
```diff
@@ -0,0 +1,70 @@
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] class FixedSizePartitionHub[T](
```
Could this be general purpose in streams instead of custom special stage in artery?

patriknw replied:
well, it's pretty special, doesn't allow some streams to terminate. It's also using Agrona so can't because of dependency.

johanandren replied:
Alright, more thinking out aloud than having strong opinions of it, didn't think of the argon dep.

johanandren replied:
Thank you, autocorrect...

johanandren reviewed (commented):
I like!

patriknw commented:
I like `statefulSink` naming

patriknw commented:
Completed this, including javadsl and docs. Ready for final review.

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Hub.scala:
```diff
@@ -91,3 +93,99 @@ object BroadcastHub {
+
+  def of[T](clazz: Class[T], partitioner: BiFunction[Integer, T, Integer], startAfterNbrOfConsumers: Int): Sink[T, Source[T, NotUsed]] =
+    of(clazz, partitioner, startAfterNbrOfConsumers, akka.stream.scaladsl.PartitionHub.defaultBufferSize)
+}
```
verified that the types (primitives and such) are right:
```
  public <T> akka.stream.javadsl.Sink<T, akka.stream.javadsl.Source<T, akka.NotUsed>> ofStateful(java.lang.Class<T>, java.util.function.Supplier<java.util.function.BiFunction<long[], T, java.lang.Long>>, int, int);
  public <T> akka.stream.javadsl.Sink<T, akka.stream.javadsl.Source<T, akka.NotUsed>> ofStateful(java.lang.Class<T>, java.util.function.Supplier<java.util.function.BiFunction<long[], T, java.lang.Long>>, int);
  public <T> akka.stream.javadsl.Sink<T, akka.stream.javadsl.Source<T, akka.NotUsed>> of(java.lang.Class<T>, java.util.function.BiFunction<java.lang.Integer, T, java.lang.Integer>, int, int);
  public <T> akka.stream.javadsl.Sink<T, akka.stream.javadsl.Source<T, akka.NotUsed>> of(java.lang.Class<T>, java.util.function.BiFunction<java.lang.Integer, T, java.lang.Integer>, int);
```

manonthegithub commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -693,3 +699,470 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+/**
+ * INTERNAL API
+ */
+private[akka] class PartitionHub[T](partitioner: () ⇒ (Array[Long], T) ⇒ Long, startAfterNbrOfConsumers: Int, bufferSize: Int)
```
There can possibly be one more scenario that is not covered. If I want to partition to the fastest consumer or only to consumers that have already demanded element, i.e. send to those consumers that are active, I guess I won't be able to do it with this partitioner. Do you think it is useful?

patriknw replied:
Good point, that would be useful. I think we can expose the queue sizes for each consumer, which should be a good indicator. I'll look into that tomorrow.

manonthegithub commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -693,3 +699,470 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+        case UnRegister(id) ⇒
+          consumers = consumers.filterNot(_.id == id)
+          consumerIds = consumerIds.filterNot(_ == id)
+          queue.remove(id)
```
All the elements that were sent to this consumer's queue will be lost?
Perhaps, it can be better to rearrange them among consumers that are left? Or it can possibly be configured whether to rearrange or just drop them?

patriknw replied:
That is no different than if the elements were in flight further downstream.
Rearrange would be weird since element order would be violated.

johanandren commented on akka-docs/src/test/java/jdocs/stream/HubDocTest.java:
```diff
@@ -137,4 +139,91 @@ public void mergeBroadcastCombination() {
+      producer.toMat(PartitionHub.of(
+          String.class, 
+          (size, elem) -> Math.abs(elem.hashCode()) % size,
+          2, 256), Keep.right());
```
Something fishy with indentation here, 2 and 256 are parameters to `of` but then it ends, could we indent the different parameter lists with different levels to make it easier to read?

Something like:
```
    producer.toMat(
      PartitionHub.of(
        String.class, 
        (size, elem) -> Math.abs(elem.hashCode()) % size,
        2, // startAfterNbrOfConsumers
        256), // bufferSize
      Keep.right());
```

patriknw replied:
significant whitespace :)

johanandren commented on akka-docs/src/test/scala/docs/stream/HubsDocSpec.scala:
```diff
@@ -104,6 +104,65 @@ class HubsDocSpec extends AkkaSpec with CompileOnlySpec {
+      // value to the left is used)
+      val runnableGraph: RunnableGraph[Source[String, NotUsed]] =
+        producer.toMat(PartitionHub.sink((size, elem) => math.abs(elem.hashCode) % size,
+          startAfterNbrOfConsumers = 2, bufferSize = 256))(Keep.right)
```
Here it's ok because of the named params.

johanandren commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala:
```diff
@@ -582,7 +608,7 @@ class HubSpec extends StreamSpec {
 
     "properly singal error to consumers arriving after producer finished" in assertAllStagesStopped {
```
signal

johanandren commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala:
```diff
@@ -370,37 +371,43 @@ class HubSpec extends StreamSpec {
   "PartitionHub" must {
```
Do we also need test coverage for:
* when partition function throws
* when materializer is abruptly terminated? 
* subscribers could actually even be materialized with separate materialisers from the one that materialized the hub, not sure if that is important to test as well?

johanandren commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -859,13 +906,15 @@ object PartitionHub {
  * INTERNAL API
  */
-private[akka] class PartitionHub[T](partitioner: () ⇒ (Array[Long], T) ⇒ Long, startAfterNbrOfStreams: Int, bufferSize: Int)
+private[akka] class PartitionHub[T](partitioner: () ⇒ (Array[Long], T) ⇒ Long, startAfterNbrOfConsumers: Int, bufferSize: Int)
```
Some fail-fast requires here on non negative buffer size and maybe nbr of consumers?

johanandren reviewed (commented):
Some comments, haven't in-depth thought about the actual stage logic yet.

manonthegithub commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -693,3 +699,470 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+      override val shape: SourceShape[T] = SourceShape(out)
+
+      override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with OutHandler {
+        private val id = idCounter.getAndIncrement()
```
I was thinking about scenarios where you want to have partitioning with priority. Then ids or priorities should be somehow specified from outside or returned, or when I know to what consumer exactly I want to send the message, I need somehow be able to get his id in advance. One possible solution is to set materialized value for the consumer source returning his id. What do you think? Does it make sense?

patriknw replied:
Materialized value of the identifier doesn't work since it's the function that needs it, so it's too late (unless you set it into some mutable variable, but that would be weird).

The identifier is a counter that is incremented in the same order as the consumers are attached/materialized. Perhaps that is enough?

manonthegithub replied:
If you think it will be useful and it is worth to add, then when Sink materializes it can return custom impl of source, which may expose something like `PrioritySource.setPriorityForFollowingMaterializations(i: Int)` which will set priority value to Consumers which will be materialized after this invocation, those priorities can be exposed in ConsumerInfo too.

patriknw replied:
I think the API complexity would be too big.

I think we have some freedom to evolve the current API if more things are needed, e.g. more accessors can be added to `ConsumerInfo`. Additional factory methods can be added.

patriknw commented:
@manonthegithub @johanandren I have exposed the consumer queue sizes, please review c89b9a6

manonthegithub commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -780,7 +781,30 @@ object PartitionHub {
-    statefulSink(() ⇒ (ids, elem) ⇒ ids(partitioner(ids.length, elem)), startAfterNbrOfConsumers, bufferSize)
+    statefulSink(() ⇒ (info, elem) ⇒ info.consumerIds(partitioner(info.size, elem)), startAfterNbrOfConsumers, bufferSize)
+
+  @DoNotInherit trait ConsumerInfo extends akka.stream.javadsl.PartitionHub.ConsumerInfo {
```
Great!).

patriknw committed:
Better use of Artery lanes for ActorSelection, #23150

* Looked into the alternativies described in the ticket, but
  they were complicated so ended up with simply including the
  uid of the sending system in the hash for selecting inbound
  lane. That should be good enough, until we have any real demand
  for something else.
* This means that different lanes can be used ActorSelection messages
  from different sending systems, i.e. good in a cluster, but same lane
  will be used for all messages originating from the same system.
* Added possibility to run the benchmarks with ActorSelection
* Added ActorSelection to the send consistency test

patriknw commented:
rebased and squashed

patriknw commented:
something with mima, I'll take a look

ktoso commented on akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala:
```diff
@@ -635,7 +635,7 @@ import scala.util.control.NonFatal
       builder
         .append("    ")
-        .append(connection.toString)
+        .append(if (connection == null) "null" else connection.toString)
```
nice catch

jrudolph commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -693,3 +700,533 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+    /**
+     * Identifiers of current consumers
+     */
+    def consumerIds: Array[Long]
```
It seems the `consumerIds` is only materialized for the purpose of the partitioner. However, like in the case of `sink` not all use cases require an array here. Putting the array in the API means that we can never change the implementation later not to provide that array.

How about providing a method `def consumerIdByIdx(idx: Int): Long` and a method `def consumerIds: immutable.IndexedSeq[Long]` (that e.g. can be implemented by `ConsumerInfoImpl` directly). The reason is to provide both is the convience to be able to use all the methods of `IndexedSeq` quickly and still have a way to access an id without having to box the Long (which would be used in `sink` above).

ktoso replied:
Good ideas, what should we leave JavaDSL though - List or Array?

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Hub.scala:
```diff
@@ -91,3 +94,122 @@ object BroadcastHub {
+  def of[T](clazz: Class[T], partitioner: BiFunction[Integer, T, Integer], startAfterNbrOfConsumers: Int): Sink[T, Source[T, NotUsed]] =
+    of(clazz, partitioner, startAfterNbrOfConsumers, akka.stream.scaladsl.PartitionHub.defaultBufferSize)
+
+  @DoNotInherit trait ConsumerInfo {
```
yeah good

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Hub.scala:
```diff
@@ -91,3 +94,122 @@ object BroadcastHub {
+    /**
+     * Identifiers of current consumers
+     */
+    def consumerIds: Array[Long]
```
I guess the Array may be ok in return type here since this is intended power user api hmm

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Hub.scala:
```diff
@@ -91,3 +94,122 @@ object BroadcastHub {
+     * Note that this is a moving target since the elements are
+     * consumed concurrently.
+     */
+    def queueSize(consumerId: Long): Int
```
nice

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Hub.scala:
```diff
@@ -91,3 +94,122 @@ object BroadcastHub {
+   * @param bufferSize Total number of elements that can be buffered. If this buffer is full, the producer
+   *   is backpressured.
+   */
+  def ofStateful[T](clazz: Class[T], partitioner: Supplier[BiFunction[ConsumerInfo, T, java.lang.Long]],
```
perhaps go with the same naming as used in the Scala one? 
`statefulSink` reads ok even with Java, even if the `of` would have been nice, the `.sink(String.class` also reads ok hm..

ktoso replied:
Yeah, there though it did only have the `of` not a stateful one... I guess it does not matter that much, the meaning makes sense

jrudolph reviewed (commented):
Great stuff. My main quibble is with the array-based partitioner API and whether (or why) the double queue design is warranted.

ktoso reviewed (commented):
Had some API questions I wanted to ask before we're stuck with it :) Otherwise looks great I think.

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -693,3 +700,533 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+    }
+
+    class PartitionQueueImpl extends PartitionQueue {
+      private val queues1 = new AtomicReferenceArray[ConsumerQueue](FixedQueues)
```
A design goal is to be able to use this in a allocation free mode in Artery, so that's why the queue impl is pluggable. That is most important for me.

The reason for the two implementations here is that the common case (as I see it) is to have a few (< 100) consumers over the lifetime of the hub but we must of course also support more.

Finding one single implementation would of course be better, but the difference ConcurrentHashMap  2.6 M/s vs FixedQueues 4.8 M/s (+ allocation rate consideration) might be too big difference to just pick the ConcurrentHashMap.

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -693,3 +700,533 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+        if (id < FixedQueues) offer1() else offer2()
+      }
+
+      override def poll(id: Long): AnyRef = {
```
I don't understand the suggestion.

patriknw replied:
note that there will be at most one concurrent consumer and one concurrent producer for each consumer queue so I don't think the contention is that problematic. ConcurrentHashMap might stripe several to the same lock, but I assume `AtomicReferenceArray` is per element.

patriknw reviewed (commented):
Thanks for great review. We agree on the API adjustments. I will not be able to update during vacation but feel free to amend if you want to merge it.

ktoso commented:
I'll apply the API changes we agreed on and so we can get this merged later today hopefully

ktoso commented on akka-docs/src/test/java/jdocs/stream/HubDocTest.java:
```diff
@@ -137,4 +143,136 @@ public void mergeBroadcastCombination() {
+  //#partition-hub-stateful-function
+  // Using a class since variable must otherwise be final.
+  // New instance is created for each materialization of the PartitionHub.
+  static class RoundRobin<T> implements BiFunction<PartitionHub.ConsumerInfo, T, Long> {
```
I dee, let's go with Long 👍

ktoso commented on akka-docs/src/test/java/jdocs/stream/HubDocTest.java:
```diff
@@ -137,4 +143,136 @@ public void mergeBroadcastCombination() {
+      producer.toMat(
+        PartitionHub.ofStateful(
+          String.class,
+          () -> new RoundRobin<String>(),
```
The inline lambda would still work with a custom interface, but can keep it as is (ToLong), not a big issue

ktoso commented on akka-docs/src/test/scala/docs/stream/HubsDocSpec.scala:
```diff
@@ -104,6 +105,86 @@ class HubsDocSpec extends AkkaSpec with CompileOnlySpec {
+      val runnableGraph: RunnableGraph[Source[String, NotUsed]] =
+        producer.toMat(PartitionHub.sink(
+          (size, elem) => math.abs(elem.hashCode) % size,
+          startAfterNbrOfConsumers = 2, bufferSize = 256))(Keep.right)
```
ok

ktoso commented on akka-docs/src/test/scala/docs/stream/HubsDocSpec.scala:
```diff
@@ -104,6 +105,86 @@ class HubsDocSpec extends AkkaSpec with CompileOnlySpec {
+      // New instance of the partitioner function and its state is created
+      // for each materialization of the PartitionHub.
+      def roundRobin(): (PartitionHub.ConsumerInfo, String) ⇒ Long = {
+        var i = -1L
```
yeah ;) long is fine then..

ktoso commented:
I pushed the slight API changes - need to head out for a while now, will be back and take a closer look if API docs and everything makes sense now.

ktoso commented:
Pushed API cleanup, please review @jrudolph

ktoso commented:
Missed to update java files, my bad, fixing

jrudolph commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -987,24 +980,36 @@ private[akka] class PartitionHub[T](
 
     private var callbackCount = 0L
 
-    private class ConsumerInfoImpl(override val consumerIds: Array[Long], val consumers: Vector[Consumer])
```
Good to be able to get rid of the parameter.

jrudolph commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -741,15 +744,15 @@ object PartitionHub {
    *   is backpressured.
    */
-  def statefulSink[T](partitioner: () ⇒ (ConsumerInfo, T) ⇒ Long, startAfterNbrOfConsumers: Int,
+  def statefulSink[T](partitioner: () ⇒ (ConsumerInfo, T) ⇒ Long, startAfterNrOfConsumers: Int,
```
In general, do we feel well enough not to introduce this as `@ApiMayChange` and keep it like this for a while until people have tested it?

patriknw replied:
@ktoso what do you think about this? I don't mind introducing it as `ApiMayChange`

jrudolph commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -987,24 +980,36 @@ private[akka] class PartitionHub[T](
+        consumers(idx).id
+
+      override def getConsumerIds: java.util.List[Long] =
+        new util.AbstractList[Long] {
```
`JavaConverters.seqAsJavaList` seems to do basically the same but this one is fine for me as well.

jrudolph commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -1073,11 +1078,9 @@ private[akka] class PartitionHub[T](
           state.getAndSet(noRegistrationsState).asInstanceOf[Open].registrations foreach { consumer ⇒
             val newConsumers = (consumerInfo.consumers :+ consumer).sortBy(_.id)
-            val newConsumerIds = consumerInfo.consumerIds :+ consumer.id
-            Arrays.sort(newConsumerIds)
```
Got a bit simpler now, even. I wonder if not sorting the ids now will have any consequences?

jrudolph reviewed (approved):
LGTM.

I wonder if we should add `@ApiMayChange` for a while to have a bit more chances to adapt APIs when people start to use it?

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -773,37 +776,27 @@ object PartitionHub {
-    /**
-     * Number of attached consumers.
-     */
-    def size: Int
```
I think it's better to have all methods in both scaladsl and javadsl even though it adds some duplication.
It makes it easier to see the full api, and users don't have to care about the implementation detail that we scaladsl extends javadsl here.

patriknw replied:
queueSize was also removed

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -987,24 +980,36 @@ private[akka] class PartitionHub[T](
         queue.size(consumerId)
 
-      override def size: Int = consumerIds.length
+      override def size: Int = consumers.length
```
`.size` since it's not an array any more

patriknw reviewed (commented):
changes are looking good, a few small comments. I can fix them myself if we agree

patriknw commented:
@ktoso I have adjusted according to my previous comments, and added ApiMayChange

patriknw commented:
rebased/squashed, will merge after validation

ktoso commented:
> @ktoso I have adjusted according to my previous comments, and added ApiMayChange

Great, thanks - LGTM :)
